File: .claude/hooks/contract-loop-gate.py
#!/usr/bin/env python3
import json
import subprocess
import sys
import time
from pathlib import Path

ROOT = Path('/Volumes/my-data/cabloy-multirepo/cabloy-basic')
STATE_FILE = Path('/tmp/cabloy-contract-loop-gate-state.json')
AUTO_SYNC_WINDOW_SECONDS = 300

TARGET_PATTERNS = [
    ('zova/src/module/', '.ts'),
    ('zova/src/module/', '.tsx'),
    ('zova/src/module/', '.jsx'),
    ('zova/src/suite/', '.ts'),
    ('zova/src/suite/', '.tsx'),
    ('zova/src/suite/', '.jsx'),
    ('vona/src/', '.ts'),
    ('vona/src/', '.tsx'),
    ('vona/src/', '.jsx'),
]

FORWARD_PATH_MARKERS = [
    '/controller/',
    '/dto/',
    '/entity/',
]

FORWARD_CONTENT_MARKERS = [
    '@Web.',
    '@Api.field',
    '@Api.body',
    'v.openapi(',
    '@Dto<',
]

REVERSE_VONA_CONTENT_MARKERS = [
    'zova-rest-cabloy-basic-admin',
    'ZovaRender.',
    'tableActionRow(',
    'tableActionBulk(',
    'ZovaRender.field(',
    'ZovaRender.cell(',
]

REVERSE_ZOVA_PATH_MARKERS = [
    '/src/bean/',
    '/src/component/',
    '/src/.metadata/',
]

REVERSE_ZOVA_CONTENT_MARKERS = [
    "declare module 'zova-module-a-openapi'",
    'IResourceTableActionRowRecord',
    '@TableCell<',
    '@Component(',
    '@Component<',
]


def _normalize_path(value):
    if not value:
        return None
    try:
        path = Path(value)
        if not path.is_absolute():
            path = ROOT / path
        return path.resolve()
    except Exception:
        return None


def _is_code_file(path: Path):
    path_text = path.as_posix()
    for prefix, suffix in TARGET_PATTERNS:
        if prefix in path_text and path_text.endswith(suffix):
            return True
    return False


def _contains_any(text, needles):
    return any(needle in text for needle in needles)


def _path_contains_any(path_text, needles):
    return any(needle in path_text for needle in needles)


def _detect_forward(path_text, content):
    if '/vona/src/' not in path_text:
        return None
    if _path_contains_any(path_text, FORWARD_PATH_MARKERS) or _contains_any(content, FORWARD_CONTENT_MARKERS):
        return 'Backend contract source may have changed.'
    return None


def _detect_reverse(path_text, content):
    if '/vona/src/' in path_text and _contains_any(content, REVERSE_VONA_CONTENT_MARKERS):
        return 'Vona code is consuming frontend metadata or render resources.'
    if '/zova/src/' in path_text and (
        _path_contains_any(path_text, REVERSE_ZOVA_PATH_MARKERS)
        or _contains_any(content, REVERSE_ZOVA_CONTENT_MARKERS)
    ):
        return 'Frontend-owned resources or metadata may affect backend consumers.'
    return None


def _analyze(path_text, content):
    return {
        'forward_reason': _detect_forward(path_text, content),
        'reverse_reason': _detect_reverse(path_text, content),
    }


def _has_signal(result):
    return bool(result['forward_reason'] or result['reverse_reason'])


def _is_high_confidence_reverse_source(path_text):
    return '/zova/src/' in path_text and _path_contains_any(path_text, REVERSE_ZOVA_PATH_MARKERS)


def _read_text(path: Path):
    try:
        return path.read_text(encoding='utf-8')
    except Exception:
        return None


def _load_state():
    if not STATE_FILE.exists():
        return {}
    try:
        return json.loads(STATE_FILE.read_text(encoding='utf-8'))
    except Exception:
        return {}


def _save_state(state):
    try:
        STATE_FILE.write_text(json.dumps(state), encoding='utf-8')
    except Exception:
        pass


def _sync_fingerprint(path: Path):
    try:
        stat = path.stat()
        return f'{path.as_posix()}:{stat.st_mtime_ns}'
    except FileNotFoundError:
        return path.as_posix()


def _should_skip_auto_sync(path: Path):
    state = _load_state()
    fingerprint = _sync_fingerprint(path)
    entry = state.get(str(ROOT))
    if not entry:
        return False
    if entry.get('fingerprint') != fingerprint:
        return False
    if time.time() - entry.get('timestamp', 0) > AUTO_SYNC_WINDOW_SECONDS:
        return False
    return True


def _mark_auto_sync(path: Path):
    state = _load_state()
    state[str(ROOT)] = {
        'fingerprint': _sync_fingerprint(path),
        'timestamp': time.time(),
    }
    _save_state(state)


def _run_command(command):
    return subprocess.run(
        command,
        cwd=ROOT,
        capture_output=True,
        text=True,
        check=False,
    )


def _summarize_process(result):
    combined = '\n'.join(part for part in [result.stdout.strip(), result.stderr.strip()] if part)
    if not combined:
        return f'command exited with code {result.returncode}'
    lines = [line.strip() for line in combined.splitlines() if line.strip()]
    tail = ' | '.join(lines[-3:])
    return f'exit {result.returncode}: {tail}'


def _auto_sync_reverse(path: Path):
    build_result = _run_command(['npm', 'run', 'build:zova:admin'])
    if build_result.returncode != 0:
        return False, f'Auto-sync failed during `npm run build:zova:admin`: {_summarize_process(build_result)}'

    deps_result = _run_command(['npm', 'run', 'deps:vona'])
    if deps_result.returncode != 0:
        return False, f'Auto-sync failed during `npm run deps:vona`: {_summarize_process(deps_result)}'

    _mark_auto_sync(path)
    return True, 'Auto-sync ran `npm run build:zova:admin` and `npm run deps:vona` for this reverse-chain edit.'


def _build_messages(path: Path, result):
    messages = [
        "Contract-loop gate: this change may affect Cabloy's bidirectional contract loop.",
    ]
    path_text = path.as_posix()

    if result['forward_reason']:
        messages.append(
            f"Forward chain: {result['forward_reason']} If backend contract truth changed, verify the emitted OpenAPI/contract output and regenerate the frontend consumer path before considering the task done."
        )
        messages.append(
            'After forward regeneration, keep frontend follow-up thin: prefer semantic model facades and reuse the existing resource-owner when the custom API still belongs to the same resource.'
        )

    if result['reverse_reason']:
        messages.append(
            f"Reverse chain: {result['reverse_reason']} If backend tooling or backend metadata will consume this handoff, refresh generated metadata when applicable, then run `npm run build:zova:admin` and `npm run deps:vona` for the Cabloy Basic Admin path."
        )
        if _is_high_confidence_reverse_source(path_text):
            if _should_skip_auto_sync(path):
                messages.append(
                    'Auto-sync skipped because the same reverse-source edit was already synced recently in this repo.'
                )
            else:
                ok, detail = _auto_sync_reverse(path)
                messages.append(detail)
                if not ok:
                    messages.append(
                        'Please review the failure before continuing. If generated artifacts already contain the expected changes but consumers still behave stale, suspect local dependency drift before making more source edits.'
                    )
                    return ' '.join(messages)
        else:
            messages.append(
                'Auto-sync did not run because this reverse-chain signal came from the consumer side rather than a high-confidence frontend source edit.'
            )
        messages.append(
            'For Cabloy Start, apply the same reverse-chain logic but resolve the Start-specific flavor names and generated-output paths from the active Start repo before executing edition-specific steps.'
        )

    return ' '.join(messages)


def run_claude_hook():
    payload = json.load(sys.stdin)
    file_path = _normalize_path(payload.get('tool_input', {}).get('file_path'))
    if not file_path or not _is_code_file(file_path):
        return 0

    content = _read_text(file_path)
    if content is None:
        return 0

    result = _analyze(file_path.as_posix(), content)
    if not _has_signal(result):
        return 0

    message = _build_messages(file_path, result)
    print(json.dumps({
        'hookSpecificOutput': {
            'hookEventName': 'PostToolUse',
            'additionalContext': message,
        },
        'systemMessage': message,
    }))
    return 0


if __name__ == '__main__':
    raise SystemExit(run_claude_hook())

File: .claude/hooks/contract-loop-gate.ts
#!/usr/bin/env node
import { existsSync, readFileSync, statSync, writeFileSync } from 'node:fs';
import { spawnSync, type SpawnSyncReturns } from 'node:child_process';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

interface HookPayload {
  tool_input?: {
    file_path?: string;
  };
}

interface AnalysisResult {
  forwardReason: string | null;
  reverseReason: string | null;
}

interface SyncStateEntry {
  fingerprint: string;
  timestamp: number;
}

interface CommandPlan {
  args: readonly string[];
  display: string;
}

interface EditionConfig {
  id: 'basic' | 'start';
  label: string;
  reverseVonaContentMarkers: readonly string[];
  reverseAutoSyncCommands: readonly CommandPlan[];
  reverseWebBuildCommand: string;
}

type ReverseSyncOutcome =
  | { kind: 'not-applicable'; message: string }
  | { kind: 'skipped'; message: string }
  | { kind: 'success'; message: string }
  | { kind: 'failure'; message: string };

type SyncState = Record<string, SyncStateEntry>;

const SCRIPT_FILE = fileURLToPath(import.meta.url);
const ROOT = path.resolve(path.dirname(SCRIPT_FILE), '../..');
const ROOT_KEY = toPosixPath(ROOT);
const STATE_FILE = path.join(os.tmpdir(), 'cabloy-contract-loop-gate-state.json');
const AUTO_SYNC_WINDOW_SECONDS = 300;

const TARGET_PATTERNS: Array<readonly [string, string]> = [
  ['zova/src/module/', '.ts'],
  ['zova/src/module/', '.tsx'],
  ['zova/src/module/', '.jsx'],
  ['zova/src/suite/', '.ts'],
  ['zova/src/suite/', '.tsx'],
  ['zova/src/suite/', '.jsx'],
  ['vona/src/', '.ts'],
  ['vona/src/', '.tsx'],
  ['vona/src/', '.jsx'],
];

const FORWARD_PATH_MARKERS = ['/controller/', '/dto/', '/entity/'];

const FORWARD_CONTENT_MARKERS = ['@Web.', '@Api.field', '@Api.body', 'v.openapi(', '@Dto<'];

const SHARED_REVERSE_VONA_CONTENT_MARKERS = [
  'ZovaRender.',
  'tableActionRow(',
  'tableActionBulk(',
];

const REVERSE_ZOVA_PATH_MARKERS = ['/src/bean/', '/src/component/', '/src/.metadata/'];

const REVERSE_ZOVA_CONTENT_MARKERS = [
  "declare module 'zova-module-a-openapi'",
  'IResourceTableActionRowRecord',
  '@TableCell<',
  '@Component(',
  '@Component<',
];

const REVERSE_AUTO_SYNC_COMMANDS: readonly CommandPlan[] = [
  { args: ['run', 'build:zova:admin'], display: 'npm run build:zova:admin' },
  { args: ['run', 'deps:vona'], display: 'npm run deps:vona' },
];

const EDITION_CONFIGS: Record<'basic' | 'start', EditionConfig> = {
  basic: {
    id: 'basic',
    label: 'Cabloy Basic',
    reverseVonaContentMarkers: ['zova-rest-cabloy-basic-admin'],
    reverseAutoSyncCommands: REVERSE_AUTO_SYNC_COMMANDS,
    reverseWebBuildCommand: 'npm run build:zova:web',
  },
  start: {
    id: 'start',
    label: 'Cabloy Start',
    reverseVonaContentMarkers: ['zova-rest-cabloy-start-admin'],
    reverseAutoSyncCommands: REVERSE_AUTO_SYNC_COMMANDS,
    reverseWebBuildCommand: 'npm run build:zova:web',
  },
};

const ACTIVE_EDITION = resolveEdition();
const FALLBACK_REVERSE_VONA_CONTENT_MARKERS = [
  ...SHARED_REVERSE_VONA_CONTENT_MARKERS,
  ...EDITION_CONFIGS.basic.reverseVonaContentMarkers,
  ...EDITION_CONFIGS.start.reverseVonaContentMarkers,
];

function toPosixPath(value: string): string {
  return value.split(path.sep).join('/');
}

function normalizePath(value?: string): string | null {
  if (!value) return null;
  try {
    const candidate = path.isAbsolute(value) ? value : path.join(ROOT, value);
    return toPosixPath(path.resolve(candidate));
  } catch {
    return null;
  }
}

function resolveEdition(): EditionConfig | null {
  if (existsSync(path.resolve(ROOT, '__CABLOY_BASIC__'))) {
    return EDITION_CONFIGS.basic;
  }
  if (existsSync(path.resolve(ROOT, '__CABLOY_START__'))) {
    return EDITION_CONFIGS.start;
  }
  return null;
}

function getReverseVonaContentMarkers(edition: EditionConfig | null): readonly string[] {
  if (!edition) return FALLBACK_REVERSE_VONA_CONTENT_MARKERS;
  return [...SHARED_REVERSE_VONA_CONTENT_MARKERS, ...edition.reverseVonaContentMarkers];
}

function isCodeFile(filePath: string): boolean {
  return TARGET_PATTERNS.some(([prefix, suffix]) => filePath.includes(prefix) && filePath.endsWith(suffix));
}

function containsAny(text: string, needles: readonly string[]): boolean {
  return needles.some(needle => text.includes(needle));
}

function pathContainsAny(filePath: string, needles: readonly string[]): boolean {
  return needles.some(needle => filePath.includes(needle));
}

function detectForward(filePath: string, content: string): string | null {
  if (!filePath.includes('/vona/src/')) return null;
  if (pathContainsAny(filePath, FORWARD_PATH_MARKERS) || containsAny(content, FORWARD_CONTENT_MARKERS)) {
    return 'Backend contract source may have changed.';
  }
  return null;
}

function detectReverse(
  filePath: string,
  content: string,
  reverseVonaContentMarkers: readonly string[],
): string | null {
  if (filePath.includes('/vona/src/') && containsAny(content, reverseVonaContentMarkers)) {
    return 'Vona code is consuming frontend metadata or render resources.';
  }
  if (
    filePath.includes('/zova/src/') &&
    (pathContainsAny(filePath, REVERSE_ZOVA_PATH_MARKERS) ||
      containsAny(content, REVERSE_ZOVA_CONTENT_MARKERS))
  ) {
    return 'Frontend-owned resources or metadata may affect backend consumers.';
  }
  return null;
}

function analyze(filePath: string, content: string, edition: EditionConfig | null): AnalysisResult {
  return {
    forwardReason: detectForward(filePath, content),
    reverseReason: detectReverse(filePath, content, getReverseVonaContentMarkers(edition)),
  };
}

function hasSignal(result: AnalysisResult): boolean {
  return Boolean(result.forwardReason || result.reverseReason);
}

function isHighConfidenceReverseSource(filePath: string): boolean {
  return filePath.includes('/zova/src/') && pathContainsAny(filePath, REVERSE_ZOVA_PATH_MARKERS);
}

function readText(filePath: string): string | null {
  try {
    return readFileSync(filePath, 'utf8');
  } catch {
    return null;
  }
}

function loadState(): SyncState {
  if (!existsSync(STATE_FILE)) return {};
  try {
    return JSON.parse(readFileSync(STATE_FILE, 'utf8')) as SyncState;
  } catch {
    return {};
  }
}

function saveState(state: SyncState): void {
  try {
    writeFileSync(STATE_FILE, JSON.stringify(state), 'utf8');
  } catch {
    // ignore state persistence failures
  }
}

function syncFingerprint(filePath: string): string {
  try {
    const stats = statSync(filePath, { bigint: true });
    return `${filePath}:${stats.mtimeNs.toString()}`;
  } catch {
    return filePath;
  }
}

function shouldSkipAutoSync(filePath: string): boolean {
  const state = loadState();
  const fingerprint = syncFingerprint(filePath);
  const entry = state[ROOT_KEY];
  if (!entry) return false;
  if (entry.fingerprint !== fingerprint) return false;
  if (Date.now() - entry.timestamp > AUTO_SYNC_WINDOW_SECONDS * 1000) return false;
  return true;
}

function markAutoSync(filePath: string): void {
  const state = loadState();
  state[ROOT_KEY] = {
    fingerprint: syncFingerprint(filePath),
    timestamp: Date.now(),
  };
  saveState(state);
}

function runNpm(args: readonly string[]): SpawnSyncReturns<string> {
  const command = process.platform === 'win32' ? 'npm.cmd' : 'npm';
  return spawnSync(command, [...args], {
    cwd: ROOT,
    encoding: 'utf8',
  });
}

function summarizeProcess(result: SpawnSyncReturns<string>): string {
  if (result.error) {
    return result.error.message;
  }
  const combined = [result.stdout?.trim(), result.stderr?.trim()].filter(Boolean).join('\n');
  const exitCode = result.status ?? 1;
  if (!combined) {
    return `command exited with code ${exitCode}`;
  }
  const lines = combined
    .split(/\r?\n/)
    .map(line => line.trim())
    .filter(Boolean);
  const tail = lines.slice(-3).join(' | ');
  return `exit ${exitCode}: ${tail}`;
}

function autoSyncReverse(filePath: string, edition: EditionConfig): ReverseSyncOutcome {
  for (const command of edition.reverseAutoSyncCommands) {
    const result = runNpm(command.args);
    if (result.status !== 0) {
      return {
        kind: 'failure',
        message: `Auto-sync failed during \`${command.display}\`: ${summarizeProcess(result)}`,
      };
    }
  }

  markAutoSync(filePath);
  const commands = edition.reverseAutoSyncCommands.map(command => `\`${command.display}\``).join(' and ');
  return {
    kind: 'success',
    message: `Auto-sync ran ${commands} for this ${edition.label} reverse-chain edit.`,
  };
}

function buildReverseGuidance(edition: EditionConfig | null): string {
  if (!edition) {
    return 'If backend tooling or backend metadata will consume this handoff, refresh generated metadata when applicable, resolve the active Cabloy edition marker before choosing the relevant Zova build and generated-output path, and then run `npm run deps:vona`.';
  }

  const autoSyncCommands = edition.reverseAutoSyncCommands
    .map(command => `\`${command.display}\``)
    .join(', then ');
  return `If backend tooling or backend metadata will consume this handoff, refresh generated metadata when applicable, run ${autoSyncCommands}, and also run \`${edition.reverseWebBuildCommand}\` when the Web flavor is affected in this ${edition.label} repo.`;
}

function resolveReverseSyncOutcome(
  filePath: string,
  result: AnalysisResult,
  edition: EditionConfig | null,
): ReverseSyncOutcome {
  if (!result.reverseReason) {
    return {
      kind: 'not-applicable',
      message: '',
    };
  }

  if (!isHighConfidenceReverseSource(filePath)) {
    return {
      kind: 'not-applicable',
      message:
        'Auto-sync did not run because this reverse-chain signal came from the consumer side rather than a high-confidence frontend source edit.',
    };
  }

  if (shouldSkipAutoSync(filePath)) {
    return {
      kind: 'skipped',
      message: 'Auto-sync skipped because the same reverse-source edit was already synced recently in this repo.',
    };
  }

  if (!edition) {
    return {
      kind: 'not-applicable',
      message:
        'Auto-sync did not run because the active Cabloy edition marker could not be resolved for this repo.',
    };
  }

  return autoSyncReverse(filePath, edition);
}

function buildMessages(
  result: AnalysisResult,
  edition: EditionConfig | null,
  reverseSyncOutcome: ReverseSyncOutcome,
): string {
  const messages = ["Contract-loop gate: this change may affect Cabloy's bidirectional contract loop."];

  if (result.forwardReason) {
    messages.push(
      `Forward chain: ${result.forwardReason} If backend contract truth changed, verify the emitted OpenAPI/contract output and regenerate the frontend consumer path before considering the task done.`,
    );
    messages.push(
      'After forward regeneration, keep frontend follow-up thin: prefer semantic model facades and reuse the existing resource-owner when the custom API still belongs to the same resource.',
    );
  }

  if (result.reverseReason) {
    messages.push(`Reverse chain: ${result.reverseReason} ${buildReverseGuidance(edition)}`);
    if (reverseSyncOutcome.message) {
      messages.push(reverseSyncOutcome.message);
    }
    if (reverseSyncOutcome.kind === 'failure') {
      messages.push(
        'Please review the failure before continuing. If generated artifacts already contain the expected changes but consumers still behave stale, suspect local dependency drift before making more source edits.',
      );
    }
  }

  return messages.join(' ');
}

function runClaudeHook(): number {
  let payload: HookPayload;
  try {
    const raw = readFileSync(0, 'utf8');
    payload = JSON.parse(raw) as HookPayload;
  } catch {
    return 0;
  }

  const filePath = normalizePath(payload.tool_input?.file_path);
  if (!filePath || !isCodeFile(filePath)) return 0;

  const content = readText(filePath);
  if (content === null) return 0;

  const result = analyze(filePath, content, ACTIVE_EDITION);
  if (!hasSignal(result)) return 0;

  const reverseSyncOutcome = resolveReverseSyncOutcome(filePath, result, ACTIVE_EDITION);
  const message = buildMessages(result, ACTIVE_EDITION, reverseSyncOutcome);
  console.log(
    JSON.stringify({
      hookSpecificOutput: {
        hookEventName: 'PostToolUse',
        additionalContext: message,
      },
      systemMessage: message,
    }),
  );
  return 0;
}

process.exit(runClaudeHook());
